### PR TITLE
Retarget drop-in kubelet configuration dir feature to Alpha

### PR DESCRIPTION
**Pull request by sohankunkerkar**
#### What type of PR is this?
/kind feature
/sig node

#### What this PR does / why we need it:
Retarget [this](https://github.com/kubernetes/enhancements/issues/3983) feature for Alpha, which includes:
1. Fixing the existing functionality
2. Adding an e2e test

#### Does this PR introduce a user-facing change?

```release-note
Fix overriding default KubeletConfig fields in drop-in configs if not set 
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- [KEP]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/3983-drop-in-configuration
```

### CODE REVIEW

File: test/e2e_node/kubelet_config_test.go
Comment (kannon92) [inline test/e2e_node/kubelet_config_test.go:19]: You should follow similar pattern for naming features like we do for the other tests.

Comment (kannon92) [inline test/e2e_node/kubelet_config_test.go:19]:
```suggestion
var _ = SIGDescribe("Kubelet Config [NodeFeature:KubeletConfig]", func() {
```

Something like that.  We typically tag these features so its easy to search for them in the tests.

File: test/e2e_node/kubelet_config_test.go
Comment (kannon92) [inline test/e2e_node/kubelet_config_test.go:22]: What are you trying to test here?  Are all the drop-in configs relevant?  

Can we get away with just testing the `systemReserved` and drop the other fields?

File: test/e2e_node/kubelet_config_test.go
Comment (kannon92) [inline test/e2e_node/kubelet_config_test.go:77]: Is there a way to know that kubelet has picked up this configuration?  Sleeps are known to be flaky in CI.

File: test/e2e_node/kubelet_config_test.go
Comment (kannon92) [inline test/e2e_node/kubelet_config_test.go:97]:
```suggestion
func writeKubeletDropinConfigFile(filename, dropinDir string, configData interface{}) error {
```

I don't think we need to export this function since its only used here.

File: test/e2e_node/kubelet_config_test.go
Comment (sohankunkerkar) [reply test/e2e_node/kubelet_config_test.go:22]: >What are you trying to test here? Are all the drop-in configs relevant?

The idea is to test whether the merging of Kubelet config fields happens appropriately. It's good to test more fields with different data types to ensure the overriding logic works as expected.

File: test/e2e_node/kubelet_config_test.go
Comment (sohankunkerkar) [reply test/e2e_node/kubelet_config_test.go:77]: Ideally, it should pick up the moment the kubelet starts. Sleep is added to allow some breathing room.

File: test/e2e_node/kubelet_config_test.go
Comment (sohankunkerkar) [reply test/e2e_node/kubelet_config_test.go:97]: Yeah, the code is still WIP. I wanted to run this test against kubernetes CI, but thanks for the suggestion.

File: test/e2e_node/kubelet_config_test.go
Comment (haircommander) [inline test/e2e_node/kubelet_config_test.go:21]: personal nit: I'm not sure we need a context here as there aren't other Its you're grouping together

File: cmd/kubelet/app/server.go
Comment (sohankunkerkar) [inline cmd/kubelet/app/server.go:358]: Let me explain my choice of using this function instead of `loadConfigFile`. I want to provide a more detailed explanation:
While running an e2e test on changes made to the kubelet config directory, I noticed a consistent test failure. The failure was due to discrepancies in the values of the following fields:
```
FileCheckFrequency
VolumeStatsAggPeriod
EvictionPressureTransitionPeriod
Authorization.Mode
EvictionHard
```
In the case of `EvictionHard`, it makes sense because we are explicitly setting the values in `loadConfigFile` if they are empty in the kubelet config. However, for the other four fields, there was no apparent reason for them to change even when we defined them in the base config. They were being overridden by default values when not explicitly set in the drop-in configs. We need to handle this in a different way to preserve the original values if they were absent in the drop-in configs. I tried investigating the code, and it appears that `DecodeKubeletConfiguration` is responsible for setting the values for kubeletconfig. However, I am uncertain if it's worth managing that logic in that function specifically for the drop-in configs, especially for those four fields.

File: cmd/kubelet/app/server.go
Comment (benluddy) [inline cmd/kubelet/app/server.go:339]: You'll lose the distinction between absent and "present but zero-valued" here. It looks like [mergo would consider](https://github.com/kubernetes/kubernetes/blob/74fefd877f0ee9445735a756270f8eac44a96d0b/vendor/github.com/imdario/mergo/mergo.go#L36-L59) the zero values for certain types (like `metav1.Duration`) non-empty and overwrite real values.

I hacked up the fixtures for one of the unit tests in https://github.com/benluddy/kubernetes/commit/4ba879188bdc6ae55bd7110dbdc52b5e8ba1a156 to demonstrate, and it gives me:

```
--- FAIL: TestMergeKubeletConfigurations (0.00s)
    --- FAIL: TestMergeKubeletConfigurations/kubelet.conf.d_overrides_kubelet.conf (0.00s)
        server_test.go:261: 
            	Error Trace:	/var/home/benluddy/redhat/kubernetes/cmd/kubelet/app/server_test.go:261
            	Error:      	Not equal: 
            	            	expected: v1.Duration{Duration:300000000000}
            	            	actual  : v1.Duration{Duration:0}
            	            	
            	            	Diff:
            	            	--- Expected
            	            	+++ Actual
            	            	@@ -1,3 +1,3 @@
            	            	 (v1.Duration) {
            	            	- Duration: (time.Duration) 300000000000
            	            	+ Duration: (time.Duration) 0
            	            	 }
            	Test:       	TestMergeKubeletConfigurations/kubelet.conf.d_overrides_kubelet.conf
            	Messages:   	Field mismatch: SyncFrequency
```

Comment (sohankunkerkar) [reply cmd/kubelet/app/server.go]: Thanks for your input. I fixed the unit test for this feature, and it's working fine now.

File: cmd/kubelet/app/server_test.go
Comment (haircommander) [inline cmd/kubelet/app/server_test.go:235]: TBH this seems kind of expensive for a check that's quite likely to be the case. Am I missing a reason to keep this?

File: test/e2e_node/services/kubelet.go
Comment (haircommander) [inline test/e2e_node/services/kubelet.go:294]: we unconditionally set this above, so I don't think we need to check this here

File: cmd/kubelet/app/server_test.go
Comment (sohankunkerkar) [reply cmd/kubelet/app/server_test.go:235]: Actually, I changed the data type for that field to `*kubeletconfiginternal.KubeletConfiguration`

Comment (sohankunkerkar) [reply cmd/kubelet/app/server_test.go]: Fixed

File: test/e2e_node/services/kubelet.go
Comment (sohankunkerkar) [reply test/e2e_node/services/kubelet.go]: Agreed, fixed!

File: test/e2e_node/kubelet_config_dir_test.go
Comment (kannon92) [inline test/e2e_node/kubelet_config_dir_test.go:32]:
```suggestion
var _ = SIGDescribe("Kubelet Config [Slow] [Serial] [Disruptive] [NodeFeature:KubeletConfigDropInDir]", func() {
```

Since Kubelet is being restarted we should mark this as serial and disruptive.

**Comment by kannon92** [comment]
/lgtm

There may be some renaming of these tests to catagorize them correctly.  We think this can go after the test freeze.  We run these tests locally and it doesn't look the test failures occur anymore from these tests.  

We should monitor and continue to investigate after this PR merges as the serial tests are not well covered by presubmits.

/assign @mrunalp @dims

**Comment by mrunalp** [comment]
Can you squash the two test commits?

File: cmd/kubelet/app/server.go
Comment (mrunalp) [inline cmd/kubelet/app/server.go:318]: nit: drop `initially`

**Review by mrunalp — APPROVED** [review]

**Comment by dims** [comment]
/approve
/lgtm

the test and go mod changes are good

**Comment by dims** [comment]
/milestone v1.29

applying milestone as this has exception from the release team here:
https://groups.google.com/g/kubernetes-sig-release/c/uvzNhRarf6w/m/sETDPDaIAAAJ

**Pull request merged**
